ffi/src/lib.rs

```diff
@@ -936,16 +958,52 @@ mod tests {
         }
     }
 
-    #[test]
-    fn engine_builder() {
-        let path = "s3://doesntmatter/foo";
+    fn get_default_engine() -> Handle<SharedExternEngine> {
+        let path = "memory:///doesntmatter/foo";
         let path = kernel_string_slice!(path);
         let builder = unsafe { ok_or_panic(get_engine_builder(path, allocate_err)) };
-        // TODO: When miri supports epoll_wait
-        // let engine = unsafe { builder_build(builder) };
+        unsafe { ok_or_panic(builder_build(builder)) }
+    }
+
+    #[test]
+    fn engine_builder() {
+        let engine = get_default_engine();
+        unsafe {
+            free_engine(engine);
+        }
+    }
 
-        // for now just rebox so it gets dropped and miri doesn't complain
-        let _box = unsafe { Box::from_raw(builder) };
+    #[tokio::test]
+    async fn test_snapshot() -> Result<(), Box<dyn std::error::Error>> {
+        let storage = Arc::new(InMemory::new());
+        add_commit(
+            storage.as_ref(),
+            0,
+            actions_to_string(vec![TestAction::Metadata]),
+        )
+        .await?;
+        let engine = DefaultEngine::new(
+            storage.clone(),
+            Path::from("/"),
+            Arc::new(TokioBackgroundExecutor::new()),
+        );
+        let engine = engine_to_handle(Arc::new(engine), allocate_err);
+        let path = "memory:///";
+
+        let snapshot =
+            unsafe { ok_or_panic(snapshot(kernel_string_slice!(path), engine.clone_ptr())) };
+
+        let version = unsafe { version(snapshot.clone_ptr()) };
+        assert_eq!(version, 0);
+
+        let table_root = unsafe { snapshot_table_root(snapshot.clone_ptr(), allocate_str) };
+        assert!(table_root.is_some());
+        let s = recover_string(table_root.unwrap());
+        assert_eq!(&s, path);
+
+        unsafe { free_snapshot(snapshot) }
```

Comment:
yea and considering `NonNull` implements `Copy` this seems reasonable? we are just a pointer at that point?

Reply:
Thinking more about this -- the actual target audience of this FFI API (e.g. C++ code) doesn't have the problem because handles are trivially copyable. So it's only rust code pretending to be FFI that needs the workaround. And for rust code, I suspect it's better to have an obvious function call to alert readers about what's going on, rather than an implicit copy?

Reply:
yeah. i was working through this a bit, and I think, since we really just want it for testing atm, it makes sense to be a `shallow_copy` function call to make it really explicit